server/src/main/java/org/elasticsearch/index/engine/LuceneSyntheticSourceChangesSnapshot.java

```diff
@@ -83,7 +83,15 @@ public LuceneSyntheticSourceChangesSnapshot(
         this.maxMemorySizeInBytes = maxMemorySizeInBytes > 0 ? maxMemorySizeInBytes : 1;
         this.sourceLoader = mapperService.mappingLookup().newSourceLoader(null, SourceFieldMetrics.NOOP);
         Set<String> storedFields = sourceLoader.requiredStoredFields();
-        this.storedFieldLoader = StoredFieldLoader.create(false, storedFields);
+
+        // If more than a few ops are requested let's enforce a sequential reader. Typically, thousands of ops may be requested.
+        // Given how LuceneSyntheticSourceChangesSnapshot accesses stored field, it should always benefit from using sequential reader.
+        //
+        // A sequential reader decompresses a block eagerly, so that increasing adjacent doc ids can access stored fields without
+        // compressing on each StoredFields#document(docId) invocation. The only downside is the last few operations in the request
+        // seq_no range are at the beginning of a block, which means stored fields for many docs are being decompressed that isn't used.
```

Comment:
Is this true also for zstd?

Reply:
I think it both true for zstd or deflate, since both de-compress an entire block.